optimizely/helpers/condition.py

```diff
@@ -12,115 +12,348 @@
 
     Returns:
-      Boolean: True if the operand evaluates to False
+      Boolean:
+        - True if the operand evaluates to False
+        - False if the operand evaluates to True
+      None: if conditions is empty or condition couldn't be evaluated
+
     """
-    if len(single_condition) != 1:
+    if not len(conditions) > 0:
+      return None
+
+    result = self.evaluate(conditions[0], leaf_evaluator)
+    return None if result is None else not result
+
+  DEFAULT_OPERATOR_TYPES = [
+    ConditionOperatorTypes.AND,
+    ConditionOperatorTypes.OR,
+    ConditionOperatorTypes.NOT
+  ]
+
+  EVALUATORS_BY_OPERATOR_TYPE = {
+    ConditionOperatorTypes.AND: and_evaluator,
+    ConditionOperatorTypes.OR: or_evaluator,
+    ConditionOperatorTypes.NOT: not_evaluator
+  }
+
+  def evaluate(self, conditions, leaf_evaluator):
+    """ Top level method to evaluate conditions
+
+    Args:
+      conditions: Nested array of and/or conditions, or a single leaf condition value of any type
+                  Example: ['and', '0', ['or', '1', '2']]
+      leaf_evaluator: Function which will be called to evaluate leaf condition values
+
+    Returns:
+      Boolean: Result of evaluating the conditions using the operator rules and the leaf evaluator.
+      None: if conditions couldn't be evaluated
+
+    """
+
+    if isinstance(conditions, list):
+      if conditions[0] in self.DEFAULT_OPERATOR_TYPES:
+        return self.EVALUATORS_BY_OPERATOR_TYPE[conditions[0]](self, conditions[1:], leaf_evaluator)
+      else:
+        # assume OR when operator is not explicit
+        return self.EVALUATORS_BY_OPERATOR_TYPE[ConditionOperatorTypes.OR](self, conditions, leaf_evaluator)
+
+    leaf_condition = conditions
+    return leaf_evaluator(leaf_condition)
+
+
+class CustomAttributeConditionEvaluator(object):
+  """ Class encapsulating methods to be used in audience leaf condition evaluation. """
+
+  CUSTOM_ATTRIBUTE_CONDITION_TYPE = 'custom_attribute'
+
+  MATCH_TYPES = [
+    ConditionMatchTypes.EXACT,
+    ConditionMatchTypes.EXISTS,
+    ConditionMatchTypes.GREATER_THAN,
+    ConditionMatchTypes.LESS_THAN,
+    ConditionMatchTypes.SUBSTRING
+  ]
+
+  def __init__(self, condition_data, attributes):
+    self.condition_data = condition_data
+    self.attributes = attributes or {}
+
+  def is_finite(self, value):
+    """ Method to validate if the given value is a number and not one of NAN, INF, -INF
+
+    Args:
+      value: Value to be validated
+
+    Returns:
+      Boolean: True if value is a number and not NAN, INF or -INF else False
+    """
+    if not isinstance(value, (numbers.Integral, float)):
+      # numbers.Integral instead of int to accomodate long integer in python 2
+      return False
+
+    if isinstance(value, bool):
+      # bool is a subclass of int
       return False
 
-    return not self.evaluate(single_condition[0])
+    if math.isnan(value) or math.isinf(value):
```

Comment:
Nit: the `isinstance(value, (numbers.Integral, float))` check at the top of this function should help, but I think it might be safer to only call `isnan`/`isinf` when `isinstance(value, float)`.
```python
if isinstance(value, float):
  if math.isnan(value) or math.isinf(value):
    return False
```
This would protect us if `isnan`/`isinf` were to start throwing `TypeError` when given integers in some future version of Python. It's [not](https://docs.python.org/3.7/library/math.html#math.isinf) [documented](https://docs.python.org/3.7/library/math.html#math.isinf), but these functions already throw a `TypeError` when given non-numeric values (as determined in my Python 2.7.10 and 3.7.0 REPLs).

Reply:
@aliabbasrizvi I think that question is better suited for the conversation(s) in which we're talking about calling `is_finite` from `is_attribute_valid`

At any rate, it looks like @oakbani accepted the original suggestion in this convo, so I'm going to resolve 😬